pkg/controller/clone-controller.go

```diff
@@ -363,18 +163,71 @@ func (c *CloneController) processPvcItem(pvc *v1.PersistentVolumeClaim) error {
 	if !checkIfLabelExists(pvc, common.CDILabelKey, common.CDILabelValue) {
 		lab = map[string]string{common.CDILabelKey: common.CDILabelValue}
 	}
-	pvc, err = updatePVC(c.clientset, pvc, anno, lab)
+	pvc, err = updatePVC(cc.clientset, pvc, anno, lab)
 	if err != nil {
 		return errors.WithMessage(err, "could not update pvc %q annotation and/or label")
 	}
 	return nil
 }
 
-// forget the passed-in key for this event and optionally log a message.
-func (c *CloneController) forgetKey(key interface{}, msg string) bool {
-	if len(msg) > 0 {
-		glog.V(3).Info(msg)
+// Select only pvcs with the 'CloneRequest' annotation and that are not being processed.
+// We forget the key unless `processPvcItem` returns an error in which case the key can be
+//ProcessNextPvcItem retried.
+
+//ProcessNextPvcItem ...
+func (cc *CloneController) ProcessNextPvcItem() bool {
+	key, shutdown := cc.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer cc.queue.Done(key)
+
+	err := cc.syncPvc(key.(string))
+	if err != nil { // processPvcItem errors may not have been logged so log here
+		glog.Errorf("error processing pvc %q: %v", key, err)
+		return true
+	}
+	return cc.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: processing pvc %q completed", key))
+}
+
+func (cc *CloneController) syncPvc(key string) error {
+	pvc, exists, err := cc.pvcFromKey(key)
+	if err != nil {
+		return err
+	} else if !exists {
+		cc.podExpectations.DeleteExpectations(key)
+	}
+
+	if pvc == nil {
+		return nil
+	}
+	//check if AnnoCloneRequest annotation exists
+	if !checkPVC(pvc, AnnCloneRequest) {
+		return nil
+	}
+	//checking for CloneOf annotation indicating that the clone was already taken care of by the provisioner (smart clone).
+	if metav1.HasAnnotation(pvc.ObjectMeta, AnnCloneOf) {
+		glog.V(3).Infof("pvc annotation %q exists indicating cloning completed, skipping pvc\n", AnnCloneOf)
```

Comment:
This needs the same format as the others, which include the name and namespace of the PVC.

Reply:
```go
glog.V(2).Infof("pvc annotation %q not found, skipping pvc \"%s/%s\"\n", AnnCloneRequest, pvc.Namespace, pvc.Name)
```

Reply:
Are there other places were I need to change this? what pattern should I look for? Is it only when we need to print the PVC details?

Reply:
Loook for "skipping pvc", the test code looks for "skipping pvc "<namespace>/<pvcname>". This is to make the 'dont do anything tests' faster, if there is a line in the log about skipping the pvc, we know it works.